### PR TITLE
app_chanspy: Add 'D' option for dual-channel audio

### PR DESCRIPTION
**Pull request by mbradeen**
Adds the 'D' option to app chanspy that causes the input and output
frames of the spied channel to be interleaved in the spy output frame.
This allows the input and output of the spied channel to be decoded
separately by the receiver.

If the 'o' option is also set, the 'D' option is ignored as the
audio being spied is inherently one direction.

Fixes: #569

UserNote: The ChanSpy application now accepts the 'D' option which
will interleave the spied audio within the outgoing frames. The
purpose of this is to allow the audio to be read as a Dual channel
stream with separate incoming and outgoing audio. Setting both the
'o' option and the 'D' option and results in the 'D' option being
ignored.

### CODE REVIEW

**Comment by mbradeen** [comment]
cherry-pick-to: 21
cherry-pick-to: 20
cherry-pick-to: 18
cherry-pick-to: certified/18.9

File: apps/app_chanspy.c [hunk 1]
```diff
@@ -471,6 +478,46 @@ static int spy_generate(struct ast_channel *chan, void *data, int len, int sampl
...
+		short write_buf[samples];
+		f = ast_audiohook_read_frame_all(&csth->spy_audiohook, samples, ast_format_slin, &fr_read, &fr_write);
+		if (f) {
+			ast_frame_free(f, 0);
+			f = NULL;
+		}
+		memset(read_buf, 0, sizeof(read_buf));
+		memset(write_buf, 0, sizeof(write_buf));
+		if (fr_read) {
+			memcpy(read_buf, fr_read->data.ptr, sizeof(read_buf));
```
Comment (InterLinked1) [inline apps/app_chanspy.c:501]: Why memset right before we memcpy?
Would make more sense to have:
if fr_read memcpy else memset

File: apps/app_chanspy.c [hunk 2]
```diff
@@ -471,6 +478,46 @@ static int spy_generate(struct ast_channel *chan, void *data, int len, int sampl
...
+		}
+		if (fr_write) {
+			memcpy(write_buf, fr_write->data.ptr, sizeof(write_buf));
+		}
+
+		for (int i = 0; i < samples; i++) {
```
Comment (InterLinked1) [inline apps/app_chanspy.c:507]: You can't declare variables in the for loop pre-condition, move `i` to the top of the block.

**Review by InterLinked1 — CHANGES_REQUESTED** [review]

File: apps/app_chanspy.c [hunk 3]
```diff
@@ -245,6 +245,11 @@
 							</enum>
 						</enumlist>
 					</option>
+					<option name="D">
+						<para>Interleave the audio coming from the channel and the audio coming to the channel in
+						the output audio as a Dual channel stream, rather than mix it. Does nothing if 'o'
```
Comment (jcolp) [inline apps/app_chanspy.c:250]: dual channel

**Review by jcolp — CHANGES_REQUESTED** [review]

**Review by jcolp — APPROVED** [review]

**Pull request merged**